Check the count of cloud node for LoadBalancer service

FengyunPan opened the pull request:
If there is no available node for LoadBalancer service, the LoadBlancer
service will become unavailable, we should update service status.
This is part of #53193
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: 
xref #53193

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

thockin commented:
This LGTM overall.  Make tests pass and ping me back.

thockin commented:
BTW - it wasn't clear if the FIXES #53193, so I changed it to `xref` - OK or wrong?

thockin commented:
Assign back to me once it is LGTM'ed and ready for approval.

thockin commented:
Sorry, canned reply.

assign back to me when it passes tests and is rready to go.

FengyunPan commented:
@thockin Yeah, 'xref' is OK. Thank you.

FengyunPan commented:
/assign @bowei PTAL, thank you.

FengyunPan commented:
/assign @bowei 
PTAL, thank you.

FengyunPan commented:
/test pull-kubernetes-bazel-test

cmluciano commented:
@kubernetes/sig-network-pr-reviews

FengyunPan commented:
/assign @matchstick 
ping @MrHohn 
PTAL,thank you.

MrHohn commented on pkg/controller/service/service_controller.go:
```diff
@@ -348,6 +359,9 @@ func (s *ServiceController) ensureLoadBalancer(service *v1.Service) (*v1.LoadBal
 		return nil, err
 	}
+	if len(nodes) == 0 {
+		return nil, fmt.Errorf("There are no available nodes for LoadBalancer service %s/%s", service.Namespace, service.Name)
```
Why should we return error here? I think this should be handled at cloud provider layer, in case cloud provider implementations might have different interpretations when there is no nodes (though yeah most of them might just return an similar error)?

FengyunPan replied:
Yes, cloud provider can't ensure a valid load balancer when there are no available nodes.
Service controller just calls EnsureLoadBalancer() when nodes isn't empty.

MrHohn replied:
I still think we shouldn't return error at service controller layer. What if the cloud provider needs to cleanup resource or reconfigure LB when there is no nodes? It seems safer if cloud provider could be aware of this as well.

MrHohn replied:
> Service controller just calls EnsureLoadBalancer() when nodes isn't empty.

I'm not seeing this definition in the cloud interface spec: https://github.com/kubernetes/kubernetes/blob/1339e4cffcd343fde2458c4e1b81e364d0060bbb/pkg/cloudprovider/cloud.go#L94-L98

FengyunPan replied:
Actually, OpenStack cloud provider need non-empty nodes at v1.7. I will check other cloud provider later.
All right, I will remove this code. Thank  you.

MrHohn commented on pkg/controller/service/service_controller.go:
```diff
@@ -306,13 +315,15 @@ func (s *ServiceController) createLoadBalancerIfNeeded(key string, service *v1.S
 
 		if err := s.persistUpdate(service); err != nil {
-			return fmt.Errorf("Failed to persist updated status to apiserver, even after retries. Giving up: %v", err), notRetryable
+			return fmt.Errorf("Failed to update service %s/%s status to apiserver: %v",
```
This error message seems a bit redundant as it will be encapsulated again. Could we somehow combine it with https://github.com/kubernetes/kubernetes/pull/53196/files#diff-1f7f903e25ab8bcbc514bb1e532e997eR301?

MrHohn commented on pkg/controller/service/service_controller.go:
```diff
@@ -676,7 +686,19 @@ func (s *ServiceController) lockedUpdateLoadBalancerHosts(service *v1.Service, h
+				glog.Errorf("Failed to update the status of LoadBalancer service %s/%s", service.Namespace, service.Name)
+			}
+
+			s.eventRecorder.Eventf(service, v1.EventTypeWarning, "UnAvailableService", "There are no available nodes for LoadBalancer service %s/%s", service.Namespace, service.Name)
```
Should it be `UnAvailableLoadBalancer` instead?

MrHohn replied:
I'm a bit worried about this event will become spam when there is no nodes, as we will be emitting this for all loadbalancer services.

FengyunPan replied:
Yes, If  there is no nodes, these load balancer service should get this event.

MrHohn commented on pkg/controller/service/service_controller.go:
```diff
@@ -676,7 +686,19 @@ func (s *ServiceController) lockedUpdateLoadBalancerHosts(service *v1.Service, h
+		// If there are no available nodes for LoadBalancer service, we need to update service status.
+		if len(hosts) == 0 {
+			previousState := v1helper.LoadBalancerStatusDeepCopy(&service.Status.LoadBalancer)
+			emptyState := &v1.LoadBalancerStatus{}
```
Why should we wipe out the `status.Ingress` field? LoadBalancer might be unavailable when there is no nodes, but presumably it still occupy the external/internal IP.  By removing this field user may assume such IP has been released, which is not true.

FengyunPan replied:
Ok, so can I just add an "UnAvailableLoadBalancer" event?
Or do you have some better ideas?

MrHohn replied:
Don't have any better idea ATM, maybe just add the event for now?

FengyunPan replied:
Ok.

MrHohn reviewed (commented):
Sorry this fell of my radar.

Commit:
Check the count of cloud node for LoadBalancer service

If there is no available node for LoadBalancer service, the LoadBlancer
service will become unavailable, we should update service status.
This is part of #53193

MrHohn commented:
Thanks, this LGTM.

/lgtm

FengyunPan commented:
/retest

MrHohn commented:
/unassign @matchstick 
/assgin @thockin 
for approval.

MrHohn commented:
/assign @thockin

MrHohn commented:
Seems PR is approved.
/retest

MrHohn commented:
/test all

MrHohn commented:
@FengyunPan Could you update the release note that service controller now emits `UnAvailableLoadBalancer` event when there is no available nodes?
